### PR TITLE
8206181: ExceptionInInitializerError: improve handling of exceptions in user-provided taglets

### PR DESCRIPTION
**Pull request by pavelrappo**
Using a custom taglet may result in exceptions, not all of which are javadoc bugs. Some of these exceptions are taglet bugs.

It was noticed that an exception in a static initializer of a custom taglet looks like a javadoc bug, which it is not. The fix was to trivially catch java.lang.ExceptionInInitializerError and translate it into an error like it is already done for some other exceptions.

While working on this fix, I noticed that there were no tests that cover various exceptions that could arise while registering a taglet:

  * ClassNotFoundException
  * NoSuchMethodException
  * InstantiationException
  * InvocationTargetException

I added a test to cover all of those as well as ExceptionInInitializerError and ClassCastException, which I proactively handled similarly to ExceptionInInitializerError.

I note that a typical taglet-registration error message is quite poor and should be improved in the future. For example, it is unclear that the method that javadoc cannot find is the nullary (i.e. parameterless or no-arg) MyTaglet constructor:

    Error - Exception java.lang.NoSuchMethodException thrown while trying to register Taglet MyTaglet...

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8206181](https://bugs.openjdk.java.net/browse/JDK-8206181): ExceptionInInitializerError: improve handling of exceptions in user-provided taglets

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to 39d087cde4c3216c8f3cb9faa1cfe3470bcadca3

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk18 pull/44/head:pull/44` \
`$ git checkout pull/44`

Update a local copy of the PR: \
`$ git checkout pull/44` \
`$ git pull https://git.openjdk.java.net/jdk18 pull/44/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 44`

View PR using the GUI difftool: \
`$ git pr show -t 44`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk18/pull/44.diff">https://git.openjdk.java.net/jdk18/pull/44.diff</a>

</details>

### CODE REVIEW

File: test/langtools/jdk/javadoc/doclet/testCustomTagletRegistration/ExceptionInInitializerErrorTaglet.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import javax.lang.model.element.Element;
+
+import com.sun.source.doctree.DocTree;
+import jdk.javadoc.doclet.Taglet;
+
+public class ExceptionInInitializerErrorTaglet implements Taglet {
+
+    static {
+        if (true) {
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testCustomTagletRegistration/ExceptionInInitializerErrorTaglet.java:36]: Does the `if (true)` have any significance? 

javac will optimize it, per JLS.

Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testCustomTagletRegistration/ExceptionInInitializerErrorTaglet.java:36]:
```
$ jshell
|  Welcome to JShell -- Version 17
|  For an introduction type: /help intro

jshell> public class Whatever {
   ...> 
   ...>     static {
   ...>         throw new RuntimeException();
   ...>     }
   ...> }
   ...> 
|  Error:
|  initializer must be able to complete normally
|      static {
|      ^-------...
```

**Review by jonathan-gibbons — APPROVED** [review]
LGTM, within the spirit of the existing code.

Please file a followup bug for a more comprehensive cleanup to protect against errors in user code, like (untrusted) user-provided tablets and doclets.   While this could be loosely modeled on `javac` `ClientCodeWrapper` and `ClientCodeException`, this time round it might be reasonable to use a `Proxy`-based wrapper, to catch exceptions in user-code and then throw them in a wrapped exception for better handling at a higher level.  That suggests that going forward, failure to init a taglet should be a fatal error.  _Note:_ one potential complication for `javadoc` not found in `javac` is the hard tool-to-doclet abstraction barrier.

**Comment by pavelrappo** [comment]
> Please file a followup bug

Done: https://bugs.openjdk.java.net/browse/JDK-8279843

**Pull request closed without merging**